Local chop for CompositionalMultiphaseFVM

paveltomin opened the pull request:
Rebase PR: https://github.com/GEOS-DEV/integratedTests/pull/44

paveltomin commented on src/coreComponents/physicsSolvers/SolverBase.hpp:
```diff
@@ -784,6 +785,9 @@ class SolverBase : public ExecutableGroup
   NonlinearSolverParameters m_nonlinearSolverParameters;
 
+  /// Flag for local chop
+  integer m_localChop = 0;
```
should it be here or in NonlinearSolverParameters?

francoishamon replied:
Yes, I would move it to `NonlinearSolverParameters`. For the naming, we can maybe stick to "scaling" to follow the nomenclature used up to now in GEOS (for instance in the function `scalingForSystemSolution`).

If you agree, in `NonlinearSolverParameters`, we could have something like:
```
enum class ScalingType : integer
{
    Global, ///< Scale the Newton update with a unique scaling factor
    Local    ///< Scale the Newton update locally (modifies the Newton direction)
};
```

paveltomin commented on src/coreComponents/physicsSolvers/fluidFlow/CompositionalMultiphaseHybridFVM.cpp:
```diff
@@ -442,13 +442,13 @@ real64 CompositionalMultiphaseHybridFVM::scalingForSystemSolution( DomainPartiti
+                                                                                          ElementSubRegionBase & subRegion )
     {
-      real64 const subRegionScalingFactor =
+      auto const subRegionData =
```
will remove auto later

paveltomin commented on src/coreComponents/physicsSolvers/fluidFlow/IsothermalCompositionalMultiphaseBaseKernels.hpp:
```diff
@@ -947,11 +984,28 @@ class ScalingAndCheckingSystemSolutionKernelBase
   {
+    StackVariables() {}
+
+    StackVariables( TYPE _localMinVal, real64 _localMaxDP, real64 _localMaxDCompDens,
```
any way to make it more elegant?

paveltomin commented:
I also added some useful output for max DP, DC, DS to better observe the nonlinear behavior, only for logLevel >= 1

francoishamon reviewed (commented):
Looks very good so far

francoishamon reviewed (commented):
Looks very good so far

paveltomin commented on src/coreComponents/physicsSolvers/fluidFlow/CompositionalMultiphaseBase.cpp:
```diff
@@ -508,24 +512,23 @@ void CompositionalMultiphaseBase::updatePhaseVolumeFraction( ObjectManagerBase &
+                                                              fluid );
+
+  maxDS = MpiWrapper::max( maxDS );
+  if( maxDS > 0 )
+    GEOS_LOG_LEVEL_RANK_0( 1, getName() << ": Max DS = "<< maxDS << " in subregion " << dataGroup.getName() );
```
Sounds good idea, I was not aware of too many subRegions cases.

paveltomin commented on src/coreComponents/physicsSolvers/fluidFlow/CompositionalMultiphaseFVM.cpp:
```diff
@@ -387,14 +389,45 @@ real64 CompositionalMultiphaseFVM::scalingForSystemSolution( DomainPartition con
+  minCompDensScalFac = MpiWrapper::min( minCompDensScalFac );
+  minTempScalFac = MpiWrapper::min( minTempScalFac );
+
+  if( m_isThermal )
+  {
+    GEOS_LOG_LEVEL_RANK_0( 1, getName() + ": Max dP = " << maxDP << ", Max dCompDens = " << maxDCompDens << ", Max dT = " << maxDT << " (before scaling)" );
+  }
```
logLevel 2 starts printing matricies I think, not that much room for levels to be honest

paveltomin commented:
Just realized few things for this PR:
1. My implementation is not compatible with line search - I added a check for that and I think it is ok for now.
2. checkSystemSolution needs to be also adjusted or the logic should be changed - update first then check. @francoishamon, what do you think?

dkachuma commented on src/coreComponents/physicsSolvers/fluidFlow/CompositionalMultiphaseFVM.cpp:
```diff
@@ -361,16 +361,18 @@ real64 CompositionalMultiphaseFVM::scalingForSystemSolution( DomainPartition con
 
   string const dofKey = dofManager.getKey( viewKeyStruct::elemDofFieldString() );
   real64 scalingFactor = 1.0;
+  real64 maxDeltaPres = 0.0, maxDeltaCompDens = 0.0, maxDeltaTemp = 0.0;
```
I can't quite tell if the locally calculated pressure and temperature changes are absolute. If not I would expect `maxDeltaPres` and `maxDeltaTemp` to start as large negative values.

paveltomin replied:
```
    // compute the change in pressure
    real64 const pres = m_pressure[ei];
    real64 const absPresChange = LvArray::math::abs( m_localSolution[stack.localRow] );
    if( stack.localMaxDeltaPres < absPresChange )
    {
      stack.localMaxDeltaPres = absPresChange;
    }
```
this is how the change is computed, so it should be absolute

dkachuma commented on src/coreComponents/physicsSolvers/NonlinearSolverParameters.cpp:
```diff
@@ -157,6 +162,10 @@ void NonlinearSolverParameters::postProcessInput()
   {
     GEOS_ERROR( " timeStepIncreaseIterLimit should be smaller than timeStepDecreaseIterLimit!!" );
```
I know it's not part of this PR, but can we take the opportunity to remove the double exclamation mark here. I find it frightening.

dkachuma commented on src/coreComponents/physicsSolvers/fluidFlow/IsothermalCompositionalMultiphaseBaseKernels.hpp:
```diff
@@ -1240,28 +1415,39 @@ class SolutionCheckKernel : public ScalingAndCheckingSystemSolutionKernelBase< i
    * @param[in] compDens the component density vector
    */
   SolutionCheckKernel( integer const allowCompDensChopping,
+                       integer const localScaling,
```
If possible can we propagate `NonlinearSolverParameters::ScalingType` just to highlight the values expected by this parameter?

paveltomin replied:
Modified, please have a look

dkachuma reviewed (approved):
Looks good to me

paveltomin commented:
> Just realized few things for this PR:
> 
> 1. My implementation is not compatible with line search - I added a check for that and I think it is ok for now.
> 2. checkSystemSolution needs to be also adjusted or the logic should be changed - update first then check. @francoishamon, what do you think?

Actually, checkSystemSolution is probably ok, it is already adjusted at flow solver level to support local scaling
